invoices/invoice_expiry_watcher.go

```diff
@@ -122,6 +186,32 @@ func makeInvoiceExpiry(paymentHash lntypes.Hash,
 	case channeldb.ContractOpen:
 		return makeTimestampExpiry(paymentHash, invoice)
 
+	// If an invoice has active htlcs, we want to expire it based on block
+	// height. We only do this for hodl invoices, since regular invoices
+	// should resolve themselves automatically.
+	case channeldb.ContractAccepted:
+		if !invoice.HodlInvoice {
+			log.Debugf("Invoice in accepted state not added to "+
+				"expiry watcher: %v", paymentHash)
+
+			return nil
+		}
+
+		var minHeight uint32
+		for _, htlc := range invoice.Htlcs {
```

Comment:
Everytime we add htlcs to the invoice, we're calling `addHtlcs` which will add the new expiry heights to the watcher. This does mean that we'll have multiple entries for one invoice in the queue, but that's pretty cheap, and we'll just skip them if the invoice was canceled in a previous run. Can come back to optimize this if we want, but would rather leave that to a follow up, since it's not too critical imo.

Reply:
Yeah I would say atm it's not a concern performance-wise. And it's not related to the newly added HTLCs but the HTLCs which are already added to the invoice. For an HTLC we have three states,
- HtlcStateAccepted
- HtlcStateSettled
- HtlcStateCanceled

Since [an HTLC with `HtlcStateSettled` can only be found in an invoice that has `ContractSettled`](https://github.com/lightningnetwork/lnd/blob/master/channeldb/invoices.go#L2185), the question is what happens when one of the `invoice.Htlcs` changes from `HtlcStateAccepted` to `HtlcStateCanceled`? Do we still use its expiry to cancel our invoice or should we update our expiry height?

Reply:
Hm yeah, if a htlc is canceled then we shouldn't be using its expiry height to cancel it. I _think_ that can only happen when we [release mpp htlcs](https://github.com/lightningnetwork/lnd/blob/41c089fbf4feb3217b1e6a651142f192cbdcc8b9/invoices/invoiceregistry.go#L370) due to timeout, and then another attempt is made with a new set of htlcs which have higher expiry heights. But we should def account for that.